src/backend_model.cc

```diff
@@ -394,27 +393,140 @@ TritonModel::IndexInstances() const
 }
 
 Status
-TritonModel::RegisterInstance(
-    std::shared_ptr<TritonModelInstance>&& instance, const bool passive)
+TritonModel::PrepareInstances(
+    const inference::ModelConfig& model_config,
+    std::vector<std::shared_ptr<TritonModelInstance>>* added_instances,
+    std::vector<std::shared_ptr<TritonModelInstance>>* removed_instances)
 {
-  if (passive) {
-    bg_passive_instances_.emplace_back(std::move(instance));
-  } else {
-    bg_instances_.emplace_back(std::move(instance));
+  added_instances->clear();
+  removed_instances->clear();
+
+  std::unordered_map<
+      TritonModelInstance::Signature,
+      std::vector<std::shared_ptr<TritonModelInstance>>>
+      existing_instances = IndexInstances();
+
+  // Iterates over all the requested instances on the model config, and decides
+  // if each requested instance can reuse an existing instance or a new instance
+  // is needed.
+  for (const auto& group : model_config.instance_group()) {
+    std::vector<std::string> profile_names;
+    for (const auto& profile_name : group.profile()) {
+      profile_names.push_back(profile_name);
+    }
+    std::vector<TritonModelInstance::SecondaryDevice> secondary_devices;
+    for (const auto& secondary_device : group.secondary_devices()) {
+      secondary_devices.emplace_back(
+          inference::
+              ModelInstanceGroup_SecondaryDevice_SecondaryDeviceKind_Name(
+                  secondary_device.kind()),
+          secondary_device.device_id());
+    }
+    for (int32_t c = 0; c < group.count(); ++c) {
+      std::string instance_name{group.name() + "_" + std::to_string(c)};
+      const bool passive = group.passive();
+      std::vector<std::tuple<
+          std::string, TRITONSERVER_InstanceGroupKind, int32_t,
+          const inference::ModelRateLimiter*>>
+          instance_settings;
```

Comment:
How about a simple struct instead of tuple for readability and less usage of `x = std::get<0>, y = std::get<1>, ...`?
```cpp
struct InstanceSetting {
  // some constructor to use with emplace_back call
  std::string policy_name;
  TRITONSERVER_InstanceGroupKind kind;
  int32_t device_id;
  const inference::ModelRateLimiter* rate_limiter_config;
};

std::vector<InstanceSetting> instance_settings;
// ...

for (const auto& setting : instance_settings) {
    const TritonModelInstance::Signature signature(setting.group, setting.device_id);
    ...
}
```

Reply:
Sure. Updated on commit: [Use struct for instance setting](https://github.com/triton-inference-server/core/pull/207/commits/7d476256d620b2f5a43eba5217f81c5cfa7ffe55)